### PR TITLE
Cache library file lists and reduce pattern matching allocations

### PR DESCRIPTION
**Pull request by emgarten**
Cache library file lists and reduce pattern matching allocations

* Cache library files across projects during restore to avoid multiple disk hits.
* Parse pattern expressions once instead of during each pattern check.

LockFileCacheBuilder stores criteria so that they are created once per target graph instead of for every package.

LockFileCacheBuilder also stores content items to avoid creating the same package items for every graph.

Slash conversions have been removed in some places, they are not needed. The package folder reader and the lock file normalizes these already.

Behavior stays, this primarily moves a few helper methods around to ensure work is done once instead of multiple times.

Test | Before | After
-- | -- | --
Restore dg file   with assets files present (force) (disable parallel) | 21s | 19s
Restore with no   assets files present (force) (disable parallel) | 62s | 18s

#### Before : 6837 MB
image

#### After : 3531 MB
image

### CODE REVIEW

**Comment by emgarten** [comment]
VSTS failed due to signing

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs [hunk 1]
```diff
@@ -137,18 +138,12 @@ public class LockFileBuilder
                         previousLibrary.Sha512 != sha512 ||
                         previousLibrary.Path != path)
                     {
+                        // The existing library did not match, create a new one.
```
Comment (mishra14) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs:141]: 👍

**Review by mishra14 — APPROVED** [review]
LGTM!

File: src/NuGet.Core/NuGet.Protocol/PackagesFolder/LocalPackageInfo.cs [hunk 2]
```diff
@@ -43,9 +47,52 @@ public class LocalPackageInfo
...
 
+        /// <summary>
+        /// Package files with OPC files filtered out.
+        /// Cached to avoid reading the same files multiple times.
+        /// </summary>
+        public IReadOnlyList<string> Files => _files.Value;
+
         public override string ToString()
         {
             return Id + " " + Version + " (" + (ManifestPath ?? ZipPath) + ")";
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Protocol/PackagesFolder/LocalPackageInfo.cs:58]: i know this is old code, and if it is called pretty often, am wondering if it makes sense to use stringbuilder to append strings here?

Comment (emgarten) [reply src/NuGet.Core/NuGet.Protocol/PackagesFolder/LocalPackageInfo.cs:58]: This is really only used for debugging purposes, we don't display or use this anywhere. String builder would be better.

**Pull request merged**